mimic/model/identity_objects.py

```diff
@@ -62,17 +80,164 @@ def not_found(message, request):
+    return _identity_error_message("unauthorized", message, UNAUTHORIZED, request)
+
+
+@implementer(IEndpointTemplate, IPlugin)
+class EndpointTemplateStore(object):
+    """
+    A :obj"`EndpointTemplateStore` is an object whiich provides the
+    functionality to integrate an Endpoint Template instance. It is only
+    meant to work with a single Endpoint Template. Use :obj:`ExternalApiStore`
+    to manage multiple Endpoint Templates.
+
+    :ivar list required_mapping: list of entries in the template that are
+        required to be present for a valid template
+    :ivar list optional_mapping: list of entries in the template that are
+        optionally present in a valid template, along with a default value.
+
+
+    Note: The OpenStack documentation[0] does not specify any required
+        parameters. For this implementation, the `id`, `region`, `type`,
+        and `name` fields are required. The `name` and `type` fields
+        are used for creating an instance of the :obj:`ExternalApiStore`
+        that will enable the listing to be put into the Service Catalog;
+        the `id` and `region` fields are used by the :obj:`ExternalApiStore`
+        for managing the templates.
+
+    [0] http://developer.openstack.org/api-ref-identity-v2-ext.html
+    """
+
+    """
+    The following lists of tuples are for mapping the template defined by
+    the OpenStack documentation to the internal strucutre for several uses.
+    The tuples have the following meaning:
+
+        Index   Meaning
+        -----   ------------------------------------------------
+        0       JSON Key in the OpenStack Identity Template Spec
+        1       Attribute Name of the :obj:`EndpointTemplateStore`
+        2       default value (only for optional_mapping)
+    """
+    required_mapping = [
+        ('id', 'id_key'),
+        ('region', 'region_key'),
+        ('type', 'type_key'),
+        ('name', 'name_key'),
+    ]
+
+    optional_mapping = [
+        ('enabled', 'enabled_key', False),
+        ('publicURL', 'public_url', u""),
+        ('internalURL', 'internal_url', u""),
+        ('adminURL', 'admin_url', u""),
+        ('RAX-AUTH:tenantAlias', 'tenant_alias', "%tenant_id%"),
+        ('versionId', 'version_id', u""),
+        ('versionInfo', 'version_info', u""),
+        ('versionList', 'version_list', u"")
+    ]
+
+    tenant_mapping = [
+        ('id', 'id_key'),
+        ('region', 'region_key'),
+        ('type', 'type_key'),
+        ('publicURL', 'public_url'),
+        ('internalURL', 'internal_url'),
+        ('adminURL', 'admin_url'),
+    ]
+
+    def __init__(self, template_dict=None):
+        """
+        Create an :obj:`EndpointTemplateStore`
+
+        :param dict template_dict (optional): optional dictionary to load the
+            Endpoint Template data from. The dictionary is simply passed to the
+            deserializer.
+        """
+        self._template_data = template_dict
+        self.id_key = None
```

Comment:
I am very wary of exposing an API surface with this many `None`s in it.  If these are all values which might or might not be present, can these attributes be private?  If a caller wants to interact with an `EndpointTemplateStore`, they have to be a constant nervous wreck about testing every field for its existence before using it.

Sandi Metz has an article on the subject of nulls (albeit in Ruby, not Python) [here](http://www.sandimetz.com/blog/2014/12/19/suspicions-of-nil) which became a really great talk [here](https://www.youtube.com/watch?v=OMPfEXIlTVE); these echo my general opinion that whenever we see a `None` we should be asking: what does the caller really want out of this object?  Can we implement two versions, the "it's nothing" and "it's something" versions of that behavior, and just have two different classes, but one interface, so the caller gets uniform results?

In this case it's probably not as simple as one or the other because there are multiple optional values, but perhaps we could just hide everything but `serialize`, for example, which is uniformly a `dict{string:string}`?